test/unit/trainers/uri_trainer_test.rb

```diff
@@ -0,0 +1,122 @@
+require 'test_helper'
+
+module NewsScraper
+  module Trainer
+    class UriTrainerTest < Minitest::Test
+      def setup
+        super
+        NewsScraper::Transformers::Article.any_instance.stubs(:transform)
+        NewsScraper::Extractors::Article.any_instance.stubs(:extract)
+      end
+
+      def test_train_with_defined_scraper_pattern
+        NewsScraper::Transformers::Article.any_instance.expects(:transform)
+        NewsScraper::Extractors::Article.any_instance.expects(:extract)
+
+        capture_subprocess_io do
+          trainer = NewsScraper::Trainer::UriTrainer.new('google.ca')
+          trainer.expects(:no_scrape_defined).never
+          trainer.train
+        end
+      end
+
+      def test_train_with_no_defined_scraper_pattern
+        NewsScraper::Transformers::Article.any_instance.expects(:transform).raises(
+          NewsScraper::Transformers::ScrapePatternNotDefined.new(root_domain: 'google.ca')
+        )
+        NewsScraper::Extractors::Article.any_instance.expects(:extract).returns('extract')
+
+        capture_subprocess_io do
+          trainer = NewsScraper::Trainer::UriTrainer.new('google.ca')
+          trainer.expects(:no_scrape_defined).with('google.ca', 'extract', 'google.ca')
+          trainer.train
+        end
+      end
+
+      def test_no_scrape_defined_with_no_step_through
+        NewsScraper::CLI.expects(:confirm).returns(false)
+        NewsScraper::Trainer::PresetsSelector.any_instance.expects(:train).never
+
+        capture_subprocess_io do
+          trainer = NewsScraper::Trainer::UriTrainer.new('google.ca')
+          trainer.expects(:save_selected_presets).never
+          trainer.no_scrape_defined('google.ca', '', 'google.ca')
+        end
+      end
+
+      def test_no_scrape_defined_with_no_save
+        NewsScraper::CLI.expects(:confirm).twice.returns(true, false)
```

Comment:
and extra metaprogramming statement will fail the tests, and we'll need to update them. Not the best, but I didnt see another way around it